src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,201 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "compression/async_stream_zstd.h"
+
+#include "bytes/bytes.h"
+#include "bytes/details/io_allocation_size.h"
+#include "compression/logger.h"
+#include "likely.h"
+#include "units.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <fmt/format.h>
+
+#include <array>
+#include <zstd.h>
+#include <zstd_errors.h>
+
+namespace compression {
+[[gnu::cold]] static void throw_zstd_err(size_t rc) {
+    if (rc == ZSTD_error_memory_allocation) {
+        ss::throw_with_backtrace<std::bad_alloc>();
+    } else {
+        ss::throw_with_backtrace<std::runtime_error>(
+          fmt::format("ZSTD error:{}", ZSTD_getErrorName(rc)));
+    }
+}
+
+static void throw_if_error(size_t rc) {
+    if (unlikely(ZSTD_isError(rc))) {
+        throw_zstd_err(rc);
+    }
+}
+
+struct zstd_workspace {
+    mutex compress_mtx;
+    mutex decompress_mtx;
+
+    ZSTD_CCtx* compress_ctx;
+    ZSTD_DCtx* decompress_ctx;
+
+    std::unique_ptr<char[], ss::free_deleter> d_workspace;
+    std::unique_ptr<char[], ss::free_deleter> c_workspace;
+
+    ss::temporary_buffer<char> d_buffer;
+    ss::temporary_buffer<char> c_buffer;
+};
```

Comment:
It would definitely be ideal. My original plan was to get this PR in first then work on futurizing the paths in `v/kafka`. There are quite a lot of functions in `v/kafka` that need to be futurized so I figured it would warrant a separate PR.